pupil_src/shared_modules/annotations.py

```diff
@@ -7,239 +7,329 @@
-            self.g_pool.quickbar.remove(b)
-            self.sub_menu.elements[:] = []
-        self.buttons = []
-
-        for e_name, hotkey in self.annotation_definitions:
-            def make_fire(e_name, hotkey):
-                return lambda _ : self.fire_annotation(e_name)
-
-            def make_remove(e_name, hotkey):
-                return lambda: self.remove_annotation((e_name, hotkey))
-
-            thumb = ui.Thumb(e_name, setter=make_fire(e_name, hotkey),
-                             getter=lambda: False, label=hotkey, hotkey=hotkey)
-            self.buttons.append(thumb)
-            self.g_pool.quickbar.append(thumb)
-            self.sub_menu.append(ui.Button(e_name+" <"+hotkey+">", make_remove(e_name, hotkey)))
+        self.customize_menu()
+        self.menu.append(ui.Text_Input("_new_annotation_name", self, label="New Name"))
+        self.menu.append(ui.Text_Input("_new_annotation_hotkey", self, label="New Hotkey"))
+        self.menu.append(ui.Button("Add Annotation Type", self._add_annotation_clicked))
+        self._annotation_list_menu = ui.Growing_Menu(
+            "Annotation List - Click to Remove"
+        )
+        self.menu.append(self._annotation_list_menu)
+        self._create_initial_annotation_list()
+
+    def _create_initial_annotation_list(self):
+        for name, hotkey in self._initial_annotation_definitions:
+            self._add_annotation_definition(name, hotkey)
+
+    @abstractmethod
+    def customize_menu(self):
+        pass
 
     def deinit_ui(self):
-        self.menu.remove(self.sub_menu)
-        self.sub_menu = None
+        self._clear_button_quickbars()
         self.remove_menu()
-        if self.buttons:
-            for b in self.buttons:
-                self.g_pool.quickbar.remove(b)
-            self.buttons = []
-
-    def append_annotation(self):
-        self.annotation_definitions.append((self.new_annotation_name,
-                                            self.new_annotation_hotkey))
-        self.update_buttons()
-
-    def remove_annotation(self, annotation):
-        try:
-            self.annotation_definitions.remove(annotation)
-        except ValueError:
-            print(annotation, self.annotation_definitions)
-        self.update_buttons()
 
-    def fire_annotation(self, annotation_label):
-        t = self.g_pool.get_timestamp()
-        logger.info('"{}"@{}'.format(annotation_label, t))
-        # you may add more field to this dictionary if you want.
-        notification = {'subject': 'annotation', 'label': annotation_label,
-                        'timestamp': t, 'duration': 0.0, 'record': True}
-        self.notify_all(notification)
+    def _clear_button_quickbars(self):
+        for annotation_info in self._annotation_infos:
+            self._remove_button_quickbar(annotation_info)
+
+    @abstractmethod
+    def fire_annotation(self, annotation_definition):
+        pass
+
+    def _add_annotation_clicked(self):
```

Comment:
To me this naming suggests that you add a "clicked annotation", confused me a little at first. Would something like `_add_annotation_callback` or `_on_add_annotation` be better? Alternatively you can make changes to allow you to call  `_add_annotation_definition` directly.